### PR TITLE
First version of part.

### PR DESCRIPTION
**Pull request by kklipsch**
@bhand-mm @wheaties @logie17 This "productizes" the makefile artifactory publishing that is copy/pasta around the golang projects I've worked on.  Give it a look.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,50 @@
 # part
 Publish to artifactory as a cli
+
+## What it does
+
+Part is a simple cli that take some command line parameters (part -h for documentation) and an artifact (zip, jar, etc).  It then creates a maven pom and publishes it and the artifact to artifactory.  This allows for easy publishing of artifacts that can be found via gavc style searches via the artifactory rest api.
```
Comment (logie17) [inline README.md]: I'm not familiar with gavc - where can I learn more?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,50 @@
...
+
+## Artifactory REST API endpoint
+
+The artifactory rest api endpoint must be provided either on the command line via the -h flag or via the ARTIFACTORY_HOST parameter.  This should be the fully formed REST API endpoint, not just the host domain. (e.g. https://artifactory.mediamath.com/artifactory).
+
+## Artifactory Authentication
+
+Part allows for the publishing to artifactory via basic auth via one of the following 3 mechanisms.
```
Comment (logie17) [inline README.md]: I'm assuming we need to setup an account on artifactory somewhere? How would we do that?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,50 @@
...
+password=password
+```
+
+### Environment Variables
+
+If no filename is passed in via the -credentials flag part will look for the ARTIFACTORY_USER and ARTIFACTORY_PASSWORD environment variables for authentication.
+
+### No Auth
+
+If no filename is passed and both environment variables are not provided, part will attempt to publis without authentication.
```
Comment (logie17) [inline README.md]: s/publis/publish/

File: credentials.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	if fileName != "" && filepath.Ext(fileName) == ".json" {
+		return getCredentialsFromJSONFile(fileName)
+	}
+
+	if fileName != "" {
+		return getCredentialsFromIniFile(fileName)
+	}
+
+	if user != "" && pass != "" {
+		return &credentials{os.Getenv(userEnvVariable), os.Getenv(passwordEnvVariable)}, nil
```
Comment (logie17) [inline credentials.go]: Didn't we already call os.Getenv(userEnvVariable) and os.Getenv(passwordEnvVariable) earlier? Couldn't we just pass in user and pass?

File: put.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func isOk(status int) bool {
+	return status > 199 && status < 300
+}
+
+func url(fileName string, host string, repo string, group string, artifact string, version string) string {
+	hostEscaped := strings.TrimRight(host, "/")
+	groupEscaped := strings.Replace(group, ".", "/", -1)
+	ur := fmt.Sprintf("%s/%s/%s/%s/%s/%s", hostEscaped, repo, groupEscaped, artifact, version, filepath.Base(fileName))
```
Comment (logie17) [inline put.go]: why not just return here instead of creating an ur?

**Comment by logie17** [comment]
Looks neat, I'll use it! :+1:

File: README.md (hunk 1 above)
Comment (kklipsch) [inline README.md]: gavc is the maven coordinate system that is very popular in jvm circles.  It stands for:
- group - usually used to identify an organization (eg com.mediamath or com.mediamath.keryx)
- artifact - the actual *thing* you are publishing (eg part or keryxd)
- version - the version of artifact to publish (eg 1.2.34 or apr-15).  By convention (and on most artifact servers) any version that ends in <VERSION>-SNAPSHOT is assumed to be a development version on the way to VERSION.  On artifact servers it is usually replaced with a unique value like timestamp, but searches for <VERSION>-SNAPSHOT will return the latest one.
- classifier - extra information if multiple gav coordinates are not enough for a single publish event.  Typically used for things like target platform (e.g. darwinamd64) or runtime (jdk14, jdk15).

[Maven GAVC](https://maven.apache.org/pom.html)
[Artifactory GAVC searches](https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-GAVCSearch)

Will update the readme to be more clear.

File: README.md (hunk 2 above)
Comment (kklipsch) [inline README.md]: artifactory will respond to ldap user/pass settings.  On the build server (and maybe locally) there are files that contain the jenkins credentials.  The builds I've done, assume a file in $HOME/.ivy2/credentials/artifactory.mediamath.com with the credentials.  So I'd check to see if you have that.

Comment (bhand-mm) [inline README.md]: Why is using Ivy creds not one of the options?  We have preferred to have functional defaults in other areas.

File: credentials.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+func getCredentials(fileName string) (*credentials, error) {
+	user := os.Getenv(userEnvVariable)
+	pass := os.Getenv(passwordEnvVariable)
+
+	if fileName != "" && filepath.Ext(fileName) == ".json" {
+		return getCredentialsFromJSONFile(fileName)
+	}
+
+	if fileName != "" {
```
Comment (bhand-mm) [inline credentials.go]: Should this check file extension as well?

File: credentials.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+type credentials struct {
+	User     string
+	Password string
+}
+
+func getCredentials(fileName string) (*credentials, error) {
+	user := os.Getenv(userEnvVariable)
+	pass := os.Getenv(passwordEnvVariable)
+
+	if fileName != "" && filepath.Ext(fileName) == ".json" {
```
Comment (bhand-mm) [inline credentials.go]: Is the result of ```filepath.Ext``` reliably lowercase?

File: credentials.go (hunk 6 above)
Comment (kklipsch) [inline credentials.go]: I mean I can check to see that it isnt .json but that is redundant given that you made it to this code path.

File: credentials.go (hunk 7 above)
Comment (kklipsch) [inline credentials.go]: if they reliably send in filenames with .json as the extension.  I treat files with the .JSON extension as ini's. YOU GOT A PROBLEM WITH THAT?

File: README.md (hunk 2 above)
Comment (kklipsch) [inline README.md]: what do you mean?  it does allow for ivy creds.

File: credentials.go (hunk 7 above)
Comment (bhand-mm) [inline credentials.go]: *protects ears*

... no, sir.

File: credentials.go (hunk 6 above)
Comment (bhand-mm) [inline credentials.go]: I suppose i was expecting you to enforce that ini's end with .ini but it seems instead that its a default interpretation of the file.

File: put.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+	return parseResponse(location, resp)
+}
+
+func isOk(status int) bool {
+	return status > 199 && status < 300
+}
+
+func url(fileName string, host string, repo string, group string, artifact string, version string) string {
+	hostEscaped := strings.TrimRight(host, "/")
```
Comment (bhand-mm) [inline put.go]: Using path.Join would eliminate the need for this.

File: put.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	groupEscaped := strings.Replace(group, ".", "/", -1)
+	return fmt.Sprintf("%s/%s/%s/%s/%s/%s", hostEscaped, repo, groupEscaped, artifact, version, filepath.Base(fileName))
```
Comment (bhand-mm) [inline put.go]: Should we net/url.QueryEscape here?

Comment (kklipsch) [inline put.go]: HOW DARE YOU

**Pull request merged**